pkg/planner/core/rule_predicate_simplification.go

```diff
@@ -147,6 +190,116 @@ func applyPredicateSimplification(sctx base.PlanContext, predicates []expression
 	return newValues
 }
 
+// Check for constant false condition.
+func unsatisfiableExpression(ctx base.PlanContext, p expression.Expression) bool {
+	if constExpr, ok := p.(*expression.Constant); ok {
+		if b, err := constExpr.Value.ToBool(ctx.GetSessionVars().StmtCtx.TypeCtx()); err == nil && b == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func unsatisfiable(ctx base.PlanContext, p1, p2 expression.Expression) bool {
+	var equalPred expression.Expression
+	var otherPred expression.Expression
+	col1, p1Type := findPredicateType(p1)
+	col2, p2Type := findPredicateType(p2)
+	if col1 != col2 || col1 == nil {
+		return false
+	}
+	if p1Type == equalPredicate {
+		equalPred = p1
+		otherPred = p2
+	} else if p2Type == equalPredicate {
+		equalPred = p2
+		otherPred = p1
+	}
+	if equalPred == nil || otherPred == nil {
+		return false
+	}
+	// Copy constant from equal predicate into other predicate.
+	equalValue := equalPred.(*expression.ScalarFunction)
+	otherValue := otherPred.(*expression.ScalarFunction)
+	newPred := expression.NewFunctionInternal(ctx.GetExprCtx(), otherValue.FuncName.L, otherValue.RetType, equalValue.GetArgs()[1], otherValue.GetArgs()[1])
+	newPredList := make([]expression.Expression, 0, 1)
+	newPredList = append(newPredList, newPred)
+	newPredList = expression.PropagateConstant(ctx.GetExprCtx(), newPredList)
+	return unsatisfiableExpression(ctx, newPredList[0])
+}
+func comparisonPred(predType predicateType) predicateType {
```

Comment:
```suggestion
}

func comparisonPred(predType predicateType) predicateType {
```

Reply:
There should be blank lines between the functions.

Reply:
OK